lightning/src/chain/channelmonitor.rs

```diff
@@ -1144,8 +1144,31 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
 
 	/// Provides a payment_hash->payment_preimage mapping. Will be automatically pruned when all
 	/// commitment_tx_infos which contain the payment hash have been revoked.
-	pub(crate) fn provide_payment_preimage(&mut self, payment_hash: &PaymentHash, payment_preimage: &PaymentPreimage) {
+	pub(crate) fn provide_payment_preimage<B: Deref, F: Deref, L: Deref>(&mut self, payment_hash: &PaymentHash, payment_preimage: &PaymentPreimage, broadcaster: &B, fee_estimator: &F, logger: &L)
+	where B::Target: BroadcasterInterface,
+		    F::Target: FeeEstimator,
+		    L::Target: Logger,
+	{
 		self.payment_preimages.insert(payment_hash.clone(), payment_preimage.clone());
+		// If the channel is force closed, try to claim the output from this preimage
+		if self.lockdown_from_offchain || self.holder_tx_signed {
+			if let Some(txid) = self.current_counterparty_commitment_txid {
+				if let Some(commitment_number) = self.counterparty_commitment_txn_on_chain.get(&txid) {
+					let (htlc_claim_reqs, set_script) = self.get_counterparty_htlc_output_claim_reqs(*commitment_number, txid, None);
+					if set_script {
+						self.counterparty_payment_script = {
+							// Note that the Network here is ignored as we immediately drop the address for the
+							// script_pubkey version
+							let payment_hash160 = WPubkeyHash::hash(&self.keys.pubkeys().payment_point.serialize());
+							Builder::new().push_opcode(opcodes::all::OP_PUSHBYTES_0).push_slice(&payment_hash160[..]).into_script()
```

Comment:
Why this code stub can't stay inside your new wrapper `get_counterparty_htlc_output_claim_reqs` ?

Reply:
Because that forces the function to borrow` &mut self` instead of `&self`, and since its parent function also borrows `&mut self` you can't have two mutable borrows.